src/org/opensolaris/opengrok/configuration/messages/Message.java

```diff
@@ -260,135 +176,238 @@ public boolean equals(Object obj) {
+    }
+
+    public enum DeliveryStatus {
+
+        OK(0x1), OVER_LIMIT(0x2), ERROR(0x4);
+
+        private final int statusCode;
+
+        DeliveryStatus(final int statusCode) {
+            this.statusCode = statusCode;
+        }
+
+        public static DeliveryStatus fromStatusCode(final int statusCode) {
+            for (DeliveryStatus ds : values()) {
+                if (ds.statusCode == statusCode) {
+                    return ds;
+                }
+            }
+            throw new IllegalArgumentException("Unknown status code " + statusCode);
+        }
+
+        public int getStatusCode() {
+            return statusCode;
+        }
+    }
+
+    private static class MessageDeserializer implements JsonDeserializer<Message> {
+
+        @Override
+        public Message deserialize(
+                final JsonElement jsonElement,
+                final Type type,
+                final JsonDeserializationContext context
+        ) {
+            String classStr = jsonElement.getAsJsonObject().get("type").getAsString();
+            Class<?> cl;
+            try {
+                cl = Class.forName(classStr);
+            } catch (ClassNotFoundException e) {
+                throw new JsonParseException(e);
+            }
+            return context.deserialize(jsonElement, cl);
+        }
+    }
+
+    private static class DurationAdapter extends TypeAdapter<Duration> {
+
+        @Override
+        public void write(final JsonWriter writer, final Duration duration) throws IOException {
+            if (duration == null) {
+                writer.nullValue();
+                return;
+            }
+            writer.value(duration.toString());
+        }
+
+        @Override
+        public Duration read(final JsonReader reader) throws IOException {
+            if (reader.peek() == JsonToken.NULL) {
+                reader.nextNull();
+                return null;
+            }
+            return Duration.parse(reader.nextString());
+        }
+    }
+
+    public static class Builder<T extends Message> {
+
+        private final Class<T> messageClass;
 
+        private final Set<String> tags = new TreeSet<>();
+
+        private String text;
+
+        private String cssClass;
+
+        private Duration duration;
+
+        public Builder(final Class<T> cl) {
+            messageClass = cl;
         }
+
+        public T build() {
+            T result = createMessage(messageClass);
+            fillData(result);
+            return result;
+        }
+
+        private void fillData(final Message message) {
+            message.tags.addAll(tags);
+
+            if (message.tags.isEmpty()) {
+                Set<String> defaultTags = message.getDefaultTags();
+                if (defaultTags != null) {
+                    message.tags.addAll(defaultTags);
+                }
+            }
+
+            if (text != null) {
+                message.text = text;
+            }
+
+            if (cssClass != null) {
```

Comment:
can you rename the `css` property so it is more descriptive ?

Reply:
but that would be misleading, one could thing that can pass arbitrary css selectors into this property

Reply:
but that would be misleading, one could thing that can pass arbitrary css selectors into this property

Reply:
Do you have any suggestion? Before there was just `class`. `cssClassAddedToWebMessageWindow`? :D

Reply:
best would be to rethink this and hide it for example in message.extras (encoded as json and translated to map)

Reply:
which would be generic and wouldn't kinda full the other messages with non-sence css class property

Reply:
I can add it to `NormalMessage`. Maybe it is also used by `AbortMessage`? Not sure though. I will check.

Reply:
Yes this is just for normal and abort.